runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -459,63 +516,69 @@ end
 -- }}}
 -- Diagnostic Movements {{{
 
---- Helper function to iterate through all of the diagnostic lines
----@return table list of diagnostics
-local _iter_diagnostic_lines = function(start, finish, step, bufnr, opts, client_id)
-  if bufnr == nil then
-    bufnr = vim.api.nvim_get_current_buf()
-  end
-
+--- Helper function to find the next diagnostic relative to a position
+---@return table the next diagnostic if found
+local _next_diagnostic = function(position, search_forward, bufnr, opts, client_id)
+  position[1] = position[1] - 1
+  bufnr = bufnr or vim.api.nvim_get_current_buf()
   local wrap = if_nil(opts.wrap, true)
-
-  local search = function(search_start, search_finish, search_step)
-    for line_nr = search_start, search_finish, search_step do
-      local line_diagnostics = M.get_line_diagnostics(bufnr, line_nr, opts, client_id)
-      if line_diagnostics and not vim.tbl_isempty(line_diagnostics) then
-        return line_diagnostics
+  local line_count = vim.api.nvim_buf_line_count(bufnr)
+  for i = 0, line_count do
+    local offset = i * (search_forward and 1 or -1)
+    local line_nr = position[1] + offset
+    if line_nr < 0 or line_nr >= line_count then
+      if not wrap then
+        return
       end
+      line_nr = (line_nr + line_count) % line_count
     end
-  end
-
-  local result = search(start, finish, step)
-
-  if wrap then
-    local wrap_start, wrap_finish
-    if step == 1 then
-      wrap_start, wrap_finish = 1, start
-    else
-      wrap_start, wrap_finish = vim.api.nvim_buf_line_count(bufnr), start
-    end
-
-    if not result then
-      result = search(wrap_start, wrap_finish, step)
+    local line_diagnostics = M.get_line_diagnostics(bufnr, line_nr, opts, client_id)
+    if line_diagnostics and not vim.tbl_isempty(line_diagnostics) then
+      if search_forward then
```

Comment:
These blocks seem like a lot of duplicated code, could turn into a small function that takes a comparison operator

Reply:
@mjlbach i did a small refactor based on your feedback :slightly_smiling_face: , what do you think?